src/lib.rs

```diff
@@ -1475,6 +1478,122 @@ where
+}
+
+#[derive(Serialize, Default)]
+struct LogoutRequestParameters {
+    id_token_hint: Option<String>,
+    logout_hint: Option<LogoutHint>,
+    client_id: Option<ClientId>,
+    post_logout_redirect_uri: Option<RedirectUrl>,
+    state: Option<CsrfToken>,
+    #[serde(serialize_with = "serialize_with_join_vec")]
+    ui_locales: Vec<LanguageTag>,
+}
+
+impl From<EndSessionUrl> for LogoutRequest {
+    fn from(value: EndSessionUrl) -> Self {
+        LogoutRequest {
+            end_session_endpoint: value,
+            parameters: Default::default(),
+        }
+    }
+}
+
+impl LogoutRequest {
+    ///
+    /// Provides an ID token previously issued by this OpenID Connect Provider as a hint about
+    /// the user's identity.
+    ///
+    pub fn set_id_token_hint<AC, GC, JE, JS, JT>(
+        mut self,
+        id_token_hint: &IdToken<AC, GC, JE, JS, JT>,
+    ) -> Self
+    where
+        AC: AdditionalClaims,
+        GC: GenderClaim,
+        JE: JweContentEncryptionAlgorithm<JT>,
+        JS: JwsSigningAlgorithm<JT>,
+        JT: JsonWebKeyType,
+    {
+        self.parameters.id_token_hint = Some(id_token_hint.to_string());
+        self
+    }
+
+    ///
+    /// Provides the OpenID Connect Provider with a hint about the user's identity.
+    ///
+    /// The nature of this hint is specific to each provider.
+    ///
+    pub fn set_logout_hint(mut self, logout_hint: LogoutHint) -> Self {
+        self.parameters.logout_hint = Some(logout_hint);
+        self
+    }
+
+    ///
+    /// Provides the OpenID Connect Provider with the client identifier.
+    ///
+    /// When both this and `id_token_hint` are set, the provider must verify that
+    /// this client id matches the one used when the ID token was issued.
+    ///
+    pub fn set_client_id(mut self, client_id: ClientId) -> Self {
+        self.parameters.client_id = Some(client_id);
+        self
+    }
+
+    ///
+    /// Provides the OpenID Connect Provider with a URI to redirect to after
+    /// the logout has been performed.
+    ///
+    pub fn set_redirect_uri(mut self, redirect_uri: RedirectUrl) -> Self {
+        self.parameters.post_logout_redirect_uri = Some(redirect_uri);
+        self
+    }
+
+    ///
+    /// Specify an opaque value that the OpenID Connect Provider should pass back
+    /// to your application using the state parameter when redirecting to redirect_uri.
+    ///
+    pub fn set_state(mut self, state: CsrfToken) -> Self {
+        self.parameters.state = Some(state);
+        self
+    }
+
+    ///
+    /// Requests the preferred languages for the user interface presented by the OpenID Connect
+    /// Provider.
+    ///
+    /// Languages should be added in order of preference.
+    ///
+    pub fn add_ui_locale(mut self, ui_locale: LanguageTag) -> Self {
+        self.parameters.ui_locales.push(ui_locale);
+        self
+    }
+
+    ///
+    /// Returns the full logout URL
+    ///
+    pub fn url(self) -> Url {
```

Comment:
the spec is a little ambiguous, but it says the following:
> OpenID Providers MUST support the use of the HTTP GET and POST methods defined in [RFC 7231](https://openid.net/specs/openid-connect-rpinitiated-1_0.html#RFC7231) [RFC7231] at the Logout Endpoint. RPs MAY use the HTTP GET or POST methods to send the logout request to the OP. If using the HTTP GET method, the request parameters are serialized using URI Query String Serialization. If using the HTTP POST method, the request parameters are serialized using Form Serialization.

What I *think* this is saying (unfortunately, there are no clarifying examples), is that when using a `GET` request, the request parameters are added to the query string, and when using a `POST` request, they're part of the `application/x-www-form-urlencoded` encoded request body.

Since this function is adding the parameters to the query string, let's rename it to `http_get_url` or something. Also, the documentation for `serde_urlencoded` says that it produces an `application/x-www-form-urlencoded` string, which is the encoding for HTML forms submitted in HTTP POST bodies. I'm not certain that this is the same encoding as for query strings in every edge case. To avoid any potential encoding issues, let's build the query string the same way that the oauth2 crate does: https://github.com/ramosbugs/oauth2-rs/blob/6b21af5c04a3c03d2dfae4c5dd7199111feae778/src/lib.rs#L1139-L1165. Be sure to omit any `None` values (the current serde serialization includes them I think).

We probably don't need to add any sort of support for `POST` requests since both `GET` and `POST` requests to the logout endpoint need to happen as a browser navigation, and I think (?) the only way to make a `POST` request like that is to submit an actual HTML form element. This isn't a server-to-server endpoint, so we wouldn't want to add a method to make the POST request ourselves, which would be missing any relevant end-user browser cookies required by the OP, and which wouldn't redirect the user to the `post_logout_redirect_uri`.